File: connect.py
"""Agent Mesh — instance connect snippet.

Import this from a Claude Code REPL or run directly. Spawns a background
asyncio event loop in a thread so sync callers can fire-and-forget.
"""
from __future__ import annotations

import asyncio
import json
import os
import sys
import threading
import time
from typing import Optional

import websockets

# ---------------- defaults ----------------
INSTANCE_ID = "cc1"
NAME = "Claude 1"
PROJECT = "OPTFINDER"
BROKER_URL = "ws://localhost:8766"
# Optional shared-token auth. None or "" => no auth header sent.
MESH_TOKEN = os.environ.get("MESH_TOKEN") or None
# ------------------------------------------

_loop: Optional[asyncio.AbstractEventLoop] = None
_loop_thread: Optional[threading.Thread] = None
_ws_holder = {"ws": None, "connected": False}
_loop_started = threading.Event()
_stop_event: Optional[asyncio.Event] = None


def _start_loop():
    global _loop, _loop_thread, _stop_event

    def runner():
        global _loop, _stop_event
        _loop = asyncio.new_event_loop()
        asyncio.set_event_loop(_loop)
        _stop_event = asyncio.Event()
        _loop_started.set()
        _loop.create_task(_client_loop())
        _loop.run_forever()

    if _loop_thread is None or not _loop_thread.is_alive():
        _loop_thread = threading.Thread(target=runner, daemon=True, name="agent-mesh-loop")
        _loop_thread.start()
        _loop_started.wait(timeout=5)


def _fmt_incoming(payload: dict) -> str:
    t = payload.get("type")
    if t == "message":
        sender = payload.get("from", "?")
        to = payload.get("to", "")
        tag = "RELAY" if to and to != "you" and sender != "you" else ""
        suffix = f" ({tag})" if tag else ""
        return f"[INCOMING from {sender}{suffix}] {payload.get('text', '')}"
    if t == "memory_write":
        m = payload.get("memory", {})
        return f"[MEMORY] {m.get('key', '')} = {m.get('value', '')}  ({m.get('type', '')}, by {m.get('by', '?')})"
    if t == "memory_init":
        items = payload.get("memory", [])
        return f"[MEMORY INIT] {len(items)} entries"
    if t == "tasks_init":
        items = payload.get("tasks", [])
        return f"[TASKS INIT] {len(items)} open task(s) assigned to me"
    if t == "task_assigned":
        task = payload.get("task", {})
        return f"[TASK ASSIGNED] {task.get('id', '?')} \"{task.get('title', '')}\" (by {task.get('created_by', '?')})"
    if t == "task_completed":
        task = payload.get("task", {})
        return f"[TASK COMPLETED] {task.get('id', '?')} done by {task.get('done_by', '?')}: {task.get('result', '')}"
    if t == "backlog":
        msgs = payload.get("messages", [])
        return f"[BACKLOG] {len(msgs)} queued messages"
    if t == "approval_decision":
        verdict = "approved" if payload.get("decision") else "rejected"
        return f"[APPROVAL DECISION] {payload.get('id', '')} {verdict}: {payload.get('action', '')}"
    if t == "control":
        return f"[CONTROL] paused={payload.get('paused')}"
    return f"[{t or 'EVENT'}] {json.dumps(payload, default=str)}"


async def _client_loop():
    delay = 1.0
    while True:
        try:
            async with websockets.connect(BROKER_URL, ping_interval=20, ping_timeout=20) as ws:
                _ws_holder["ws"] = ws
                _ws_holder["connected"] = True
                # register
                reg_payload = {
                    "type": "register",
                    "id": INSTANCE_ID,
                    "name": NAME,
                    "project": PROJECT,
                }
                if MESH_TOKEN:
                    reg_payload["token"] = MESH_TOKEN
                await ws.send(json.dumps(reg_payload))
                print(f"[CONNECTED] {INSTANCE_ID} -> {BROKER_URL}")
                delay = 1.0
                async for raw in ws:
                    try:
                        payload = json.loads(raw)
                    except Exception:
                        continue
                    if payload.get("type") == "auth_failed":
                        print(f"[AUTH FAILED] {payload.get('reason', '')} — check MESH_TOKEN")
                        return
                    if payload.get("type") == "backlog":
                        msgs = payload.get("messages", [])
                        print(f"[BACKLOG] {len(msgs)} queued messages")
                        for m in msgs:
                            print("  " + _fmt_incoming(m))
                    else:
                        print(_fmt_incoming(payload))
        except (websockets.ConnectionClosed, ConnectionRefusedError, OSError) as e:
            print(f"[DISCONNECTED] {e}; retry in {delay:.1f}s")
        except Exception as e:
            print(f"[ERROR] {e}; retry in {delay:.1f}s")
        finally:
            _ws_holder["ws"] = None
            _ws_holder["connected"] = False
        await asyncio.sleep(delay)
        delay = min(delay * 2, 30.0)


async def _send_json(payload: dict):
    ws = _ws_holder.get("ws")
    if ws is None:
        print("[SKIP] not connected")
        return
    try:
        await ws.send(json.dumps(payload, default=str))
    except Exception as e:
        print(f"[SEND ERROR] {e}")


def _schedule(coro):
    _start_loop()
    if _loop is None:
        return None
    return asyncio.run_coroutine_threadsafe(coro, _loop)


# ---------------- public API ----------------
def broker_send(text: str, to: str = "you"):
    _schedule(_send_json({"type": "message", "to": to, "text": text}))


def broker_broadcast(text: str):
    _schedule(_send_json({"type": "broadcast", "text": text}))


def broker_status(task: str, workload: int = 0):
    _schedule(_send_json({"type": "status", "task": task, "workload": workload}))


def broker_approve_request(action: str, risk: str = "low", detail: str = ""):
    _schedule(_send_json({
        "type": "approval_request",
        "action": action,
        "risk": risk,
        "detail": detail,
    }))


def broker_memory(key: str, value, mem_type: str = "contract"):
    _schedule(_send_json({
        "type": "memory_write",
        "key": key,
        "value": value,
        "mem_type": mem_type,
    }))


def broker_typing(value: bool = True):
    _schedule(_send_json({"type": "typing", "value": bool(value)}))


def broker_task_create(title: str, assignee: str = "", priority: str = "normal",
                       deps: Optional[list] = None):
    """Create a task. If `assignee` is another instance id, they get task_assigned."""
    _schedule(_send_json({
        "type": "task_create",
        "title": title,
        "assignee": assignee,
        "priority": priority,
        "deps": list(deps or []),
    }))


def broker_task_claim(task_id: str):
    _schedule(_send_json({"type": "task_claim", "id": task_id}))


def broker_task_status(task_id: str, status: str):
    _schedule(_send_json({"type": "task_status", "id": task_id, "status": status}))


def broker_task_done(task_id: str, result: str = ""):
    _schedule(_send_json({"type": "task_done", "id": task_id, "result": result}))


# Start the loop on import
_start_loop()


# ---------------- CLI ----------------
def _interactive():
    print(f"Agent Mesh connect — id={INSTANCE_ID} name={NAME} project={PROJECT}")
    print("Commands: send <to> <text...> | broadcast <text...> | status <task...> [|<workload>] | quit")
    while True:
        try:
            line = input("> ").strip()
        except (EOFError, KeyboardInterrupt):
            print()
            break
        if not line:
            continue
        if line in ("quit", "exit"):
            break
        parts = line.split(" ", 2)
        cmd = parts[0].lower()
        if cmd == "send" and len(parts) >= 3:
            broker_send(parts[2], to=parts[1])
        elif cmd == "broadcast" and len(parts) >= 2:
            text = line[len("broadcast"):].strip()
            broker_broadcast(text)
        elif cmd == "status" and len(parts) >= 2:
            payload = line[len("status"):].strip()
            workload = 0
            if "|" in payload:
                payload, w = payload.rsplit("|", 1)
                try:
                    workload = int(w.strip())
                except Exception:
                    workload = 0
            broker_status(payload.strip(), workload=workload)
        else:
            print("Unknown command")


if __name__ == "__main__":
    # ensure loop is up
    _start_loop()
    time.sleep(0.5)
    _interactive()

File: clients/connect.ts
/**
 * Agent Mesh — TypeScript client.
 *
 * Mirrors the helper surface of `connect.py` so Node 22+/Bun-based Claude
 * Code instances (or external scripts) can join the mesh without a Python
 * runtime.
 *
 * Zero runtime deps — uses the global `WebSocket` available in Node 22+,
 * Bun, Deno, and browsers. For Node <22 install the `ws` package and pass
 * an explicit constructor via `options.websocketImpl`.
 */

// ---------------- types ----------------

export interface MeshMessage {
  id?: string;
  from: string;
  to: string;
  text: string;
  ts?: string;
}

export interface MeshTask {
  id: string;
  title: string;
  assignee: string;
  priority: string;
  deps: string[];
  status: string;
  created_by: string;
  ts: string;
  result?: string;
  done_by?: string;
  done_at?: string;
}

export interface MeshMemory {
  id: string;
  key: string;
  value: unknown;
  type: string;
  by: string;
  ts: string;
}

export interface MeshApproval {
  id: string;
  from: string;
  action: string;
  risk: string;
  detail: string;
  status: string;
  ts: string;
}

export interface MeshVote {
  id: string;
  question: string;
  options: string[];
  ballots: Record<string, string>;
  threshold: number | null;
  status: string;
  winner: string | null;
  created_by: string;
  ts: string;
}

// Discriminated union of every event the broker can push to an instance.
export type IncomingEvent =
  | { type: "message"; id?: string; from: string; to: string; text: string; ts?: string }
  | { type: "memory_write"; memory: MeshMemory }
  | { type: "memory_init"; memory: MeshMemory[]; ts: string }
  | { type: "tasks_init"; tasks: MeshTask[]; ts: string }
  | { type: "task_assigned"; task: MeshTask }
  | { type: "task_completed"; task: MeshTask }
  | { type: "backlog"; messages: IncomingEvent[] }
  | { type: "approval_pending"; id: string; action: string }
  | { type: "approval_decision"; id: string; action: string; decision: boolean }
  | { type: "approval_request"; approval: MeshApproval }
  | { type: "vote_pending"; vote_id: string; question: string; options: string[] }
  | { type: "vote_open"; vote: MeshVote }
  | { type: "vote_resolved"; vote_id: string; winner: string; vote: MeshVote }
  | { type: "control"; paused: boolean }
  | { type: "error"; error: string; ref?: string }
  | { type: string; [k: string]: unknown };

export type EventType = IncomingEvent["type"];
export type EventHandler<E extends IncomingEvent = IncomingEvent> = (event: E) => void;

export interface ConnectOptions {
  id: string;
  name: string;
  project: string;
  brokerUrl?: string;
  token?: string;
  /** Override the WebSocket implementation (useful on Node <22 with the `ws` package). */
  websocketImpl?: typeof WebSocket;
  /** Suppress the built-in `[CONNECTED] / [DISCONNECTED]` logging. */
  quiet?: boolean;
}

// ---------------- client ----------------

const DEFAULT_BROKER = "ws://localhost:8766";

/** Internal: a pending approveAndWait() call awaiting a broker decision. */
interface ApprovalWaiter {
  apId: string | null;
  resolve: (decision: boolean | null) => void;
  timer: ReturnType<typeof setTimeout> | null;
  settled: boolean;
}

/** Internal: a pending voteAndWait() call awaiting vote resolution. */
interface VoteWaiter {
  voteId: string | null;
  resolve: (winner: string | null) => void;
  timer: ReturnType<typeof setTimeout> | null;
  settled: boolean;
}

export class MeshClient {
  readonly id: string;
  readonly name: string;
  readonly project: string;
  readonly brokerUrl: string;
  readonly token?: string;

  private ws: WebSocket | null = null;
  private handlers = new Map<string, Set<EventHandler>>();
  private wildcardHandlers = new Set<EventHandler>();
  private wsImpl: typeof WebSocket;
  private quiet: boolean;

  private reconnectDelay = 1.0;
  private stopped = false;
  private connectedOnce = false;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;

  // approveAndWait correlation. A waiter is queued the moment we send the
  // request; the broker's `approval_pending` echo binds it to a server id
  // (FIFO), and `approval_decision` then resolves it. Mirrors connect.py.
  private approvalWaiters: ApprovalWaiter[] = [];
  private approvalById = new Map<string, ApprovalWaiter>();

  // voteAndWait correlation. `vote_pending` binds the waiter to a vote_id
  // (FIFO over pending creates); `vote_resolved` resolves it.
  private voteCreateQueue: VoteWaiter[] = [];
  private voteById = new Map<string, VoteWaiter>();

  constructor(opts: ConnectOptions) {
    this.id = opts.id;
    this.name = opts.name;
    this.project = opts.project;
    this.brokerUrl = opts.brokerUrl ?? DEFAULT_BROKER;
    this.token = opts.token;
    this.quiet = opts.quiet ?? false;

    const impl = opts.websocketImpl ?? (globalThis as { WebSocket?: typeof WebSocket }).WebSocket;
    if (!impl) {
      throw new Error(
        "No WebSocket implementation found. Use Node 22+, Bun, Deno, a browser, " +
          "or pass `websocketImpl` (e.g. `import WebSocket from 'ws'`).",
      );
    }
    this.wsImpl = impl;
  }

  // --------- lifecycle ---------

  /** Open the socket. Resolves after the first `register` payload has been
   *  flushed (i.e. the broker is aware of us). */
  async start(): Promise<void> {
    this.stopped = false;
    await this._openOnce();
  }

  /** Permanently close — no further reconnect attempts. */
  disconnect(): void {
    this.stopped = true;
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    if (this.ws) {
      try {
        this.ws.close();
      } catch {
        /* noop */
      }
      this.ws = null;
    }
    // Settle any blocking waiters so callers don't hang on a closed client.
    const approvals = [...this.approvalWaiters, ...this.approvalById.values()];
    this.approvalWaiters = [];
    this.approvalById.clear();
    for (const w of approvals) this._settleApproval(w, null);
    const votes = [...this.voteCreateQueue, ...this.voteById.values()];
    this.voteCreateQueue = [];
    this.voteById.clear();
    for (const w of votes) this._settleVote(w, null);
  }

  private _openOnce(): Promise<void> {
    return new Promise((resolve, reject) => {
      let settled = false;
      let ws: WebSocket;
      try {
        ws = new this.wsImpl(this.brokerUrl);
      } catch (err) {
        this._scheduleReconnect();
        reject(err);
        return;
      }
      this.ws = ws;

      ws.addEventListener("open", () => {
        const registerPayload: Record<string, unknown> = {
          type: "register",
          id: this.id,
          name: this.name,
          project: this.project,
        };
        if (this.token) registerPayload.token = this.token;
        try {
          ws.send(JSON.stringify(registerPayload));
          if (!this.quiet) {
            // eslint-disable-next-line no-console
            console.log(`[CONNECTED] ${this.id} -> ${this.brokerUrl}`);
          }
          this.reconnectDelay = 1.0;
          this.connectedOnce = true;
          if (!settled) {
            settled = true;
            resolve();
          }
        } catch (err) {
          if (!settled) {
            settled = true;
            reject(err);
          }
        }
      });

      ws.addEventListener("message", (evt: MessageEvent) => {
        const raw = typeof evt.data === "string" ? evt.data : evt.data?.toString?.();
        if (!raw) return;
        let payload: IncomingEvent;
        try {
          payload = JSON.parse(raw) as IncomingEvent;
        } catch {
          return;
        }
        this._dispatch(payload);
      });

      ws.addEventListener("error", () => {
        // The `close` handler runs after `error`; rely on it for reconnect.
      });

      ws.addEventListener("close", () => {
        this.ws = null;
        if (!this.quiet) {
          // eslint-disable-next-line no-console
          console.log(`[DISCONNECTED] retry in ${this.reconnectDelay.toFixed(1)}s`);
        }
        if (!settled) {
          settled = true;
          // First-attempt failure: still schedule reconnect but surface error.
          this._scheduleReconnect();
          reject(new Error("WebSocket closed before register completed"));
          return;
        }
        if (!this.stopped) {
          this._scheduleReconnect();
        }
      });
    });
  }

  private _scheduleReconnect(): void {
    if (this.stopped) return;
    if (this.reconnectTimer) return;
    const delayMs = Math.floor(this.reconnectDelay * 1000);
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this._openOnce().catch(() => {
        /* swallowed; close handler will reschedule */
      });
    }, delayMs);
    this.reconnectDelay = Math.min(this.reconnectDelay * 2, 30.0);
  }

  private _dispatch(payload: IncomingEvent): void {
    const t = payload.type;
    // Resolve any blocking approveAndWait/voteAndWait waiters first. Backlog
    // frames replay queued events, so unwrap them too (mirrors connect.py).
    if (t === "backlog" && Array.isArray((payload as { messages?: unknown[] }).messages)) {
      for (const m of (payload as { messages: IncomingEvent[] }).messages) {
        this._handlePending(m);
      }
    } else {
      this._handlePending(payload);
    }
    const set = this.handlers.get(t);
    if (set) {
      for (const h of set) {
        try {
          h(payload);
        } catch (err) {
          // eslint-disable-next-line no-console
          console.error(`[handler error type=${t}]`, err);
        }
      }
    }
    for (const h of this.wildcardHandlers) {
      try {
        h(payload);
      } catch (err) {
        // eslint-disable-next-line no-console
        console.error(`[wildcard handler error type=${t}]`, err);
      }
    }
  }

  /** Route correlation events to any blocking approveAndWait/voteAndWait calls. */
  private _handlePending(payload: IncomingEvent): void {
    switch (payload.type) {
      case "approval_pending": {
        const apId = (payload as { id?: string }).id;
        if (!apId) return;
        // Bind the oldest unbound waiter (FIFO) to this server-assigned id.
        const waiter = this.approvalWaiters.shift();
        if (waiter) {
          waiter.apId = apId;
          this.approvalById.set(apId, waiter);
        }
        return;
      }
      case "approval_decision": {
        const apId = (payload as { id?: string }).id;
        if (!apId) return;
        const waiter = this.approvalById.get(apId);
        if (!waiter) return;
        this.approvalById.delete(apId);
        this._settleApproval(waiter, Boolean((payload as { decision?: unknown }).decision));
        return;
      }
      case "vote_pending": {
        const voteId = (payload as { vote_id?: string }).vote_id;
        if (!voteId) return;
        const waiter = this.voteCreateQueue.shift();
        if (waiter) {
          waiter.voteId = voteId;
          this.voteById.set(voteId, waiter);
        }
        return;
      }
      case "vote_resolved": {
        const voteId = (payload as { vote_id?: string }).vote_id;
        if (!voteId) return;
        const waiter = this.voteById.get(voteId);
        if (!waiter) return;
        this.voteById.delete(voteId);
        const winner = (payload as { winner?: string }).winner;
        this._settleVote(waiter, typeof winner === "string" ? winner : null);
        return;
      }
      default:
        return;
    }
  }

  private _settleApproval(waiter: ApprovalWaiter, decision: boolean | null): void {
    if (waiter.settled) return;
    waiter.settled = true;
    if (waiter.timer) {
      clearTimeout(waiter.timer);
      waiter.timer = null;
    }
    waiter.resolve(decision);
  }

  private _settleVote(waiter: VoteWaiter, winner: string | null): void {
    if (waiter.settled) return;
    waiter.settled = true;
    if (waiter.timer) {
      clearTimeout(waiter.timer);
      waiter.timer = null;
    }
    waiter.resolve(winner);
  }

  // --------- subscriptions ---------

  /**
   * Register an event handler.
   *
   *   client.on("message", e => console.log(e.from, e.text));
   *   client.on("*", e => log(e));   // catch-all
   *
   * Returns an unsubscribe function.
   */
  on<T extends EventType>(
    eventType: T,
    handler: EventHandler<Extract<IncomingEvent, { type: T }>>,
  ): () => void;
  on(eventType: "*", handler: EventHandler): () => void;
  on(eventType: string, handler: EventHandler): () => void {
    if (eventType === "*") {
      this.wildcardHandlers.add(handler);
      return () => this.wildcardHandlers.delete(handler);
    }
    let set = this.handlers.get(eventType);
    if (!set) {
      set = new Set();
      this.handlers.set(eventType, set);
    }
    set.add(handler);
    return () => set!.delete(handler);
  }

  // --------- outgoing helpers (mirror connect.py) ---------

  private _sendJson(payload: Record<string, unknown>): void {
    const ws = this.ws;
    if (!ws || ws.readyState !== 1 /* OPEN */) {
      if (!this.quiet) {
        // eslint-disable-next-line no-console
        console.log("[SKIP] not connected");
      }
      return;
    }
    try {
      ws.send(JSON.stringify(payload));
    } catch (err) {
      // eslint-disable-next-line no-console
      console.error("[SEND ERROR]", err);
    }
  }

  send(text: string, to: string = "you"): void {
    this._sendJson({ type: "message", to, text });
  }

  broadcast(text: string): void {
    this._sendJson({ type: "broadcast", text });
  }

  status(task: string, workload: number = 0): void {
    this._sendJson({ type: "status", task, workload });
  }

  memory(key: string, value: unknown, memType: string = "contract"): void {
    this._sendJson({ type: "memory_write", key, value, mem_type: memType });
  }

  approveRequest(action: string, risk: string = "low", detail: string = ""): void {
    this._sendJson({ type: "approval_request", action, risk, detail });
  }

  /**
   * Send an approval_request and resolve once the broker returns a decision.
   *
   * Resolves `true` if approved, `false` if rejected, and `null` on timeout.
   * Mirrors `broker_approve_and_wait` in connect.py.
   */
  approveAndWait(
    action: string,
    risk: string = "low",
    detail: string = "",
    timeoutMs: number = 300_000,
  ): Promise<boolean | null> {
    return new Promise<boolean | null>((resolve) => {
      const waiter: ApprovalWaiter = { apId: null, resolve, timer: null, settled: false };
      // Queue before sending so an `approval_pending` echo binds this waiter.
      this.approvalWaiters.push(waiter);
      waiter.timer = setTimeout(() => {
        if (waiter.settled) return;
        // Timeout cleanup: drop from whichever map/queue still references us.
        const qi = this.approvalWaiters.indexOf(waiter);
        if (qi !== -1) this.approvalWaiters.splice(qi, 1);
        if (waiter.apId && this.approvalById.get(waiter.apId) === waiter) {
          this.approvalById.delete(waiter.apId);
        }
        this._settleApproval(waiter, null);
      }, timeoutMs);
      this._sendJson({ type: "approval_request", action, risk, detail });
    });
  }

  /**
   * Create a vote and resolve once it resolves on the broker.
   *
   * Resolves with the winning option, or `null` on timeout. Mirrors
   * `broker_vote_and_wait` in connect.py.
   */
  voteAndWait(
    question: string,
    options: string[],
    threshold: number | null = null,
    timeoutMs: number = 300_000,
  ): Promise<string | null> {
    return new Promise<string | null>((resolve) => {
      const waiter: VoteWaiter = { voteId: null, resolve, timer: null, settled: false };
      // Queue before sending so a `vote_pending` echo binds this waiter's id.
      this.voteCreateQueue.push(waiter);
      waiter.timer = setTimeout(() => {
        if (waiter.settled) return;
        const qi = this.voteCreateQueue.indexOf(waiter);
        if (qi !== -1) this.voteCreateQueue.splice(qi, 1);
        if (waiter.voteId && this.voteById.get(waiter.voteId) === waiter) {
          this.voteById.delete(waiter.voteId);
        }
        this._settleVote(waiter, null);
      }, timeoutMs);
      this._sendJson({
        type: "vote_create",
        question,
        options: [...options],
        threshold,
      });
    });
  }

  voteCreate(question: string, options: string[], threshold: number | null = null): void {
    this._sendJson({ type: "vote_create", question, options: [...options], threshold });
  }

  voteCast(voteId: string, option: string): void {
    this._sendJson({ type: "vote_cast", vote_id: voteId, option });
  }

  typing(value: boolean = true): void {
    this._sendJson({ type: "typing", value: Boolean(value) });
  }

  taskCreate(
    title: string,
    assignee: string = "",
    priority: string = "normal",
    deps: string[] = [],
  ): void {
    this._sendJson({
      type: "task_create",
      title,
      assignee,
      priority,
      deps: [...deps],
    });
  }

  taskClaim(id: string): void {
    this._sendJson({ type: "task_claim", id });
  }

  taskStatus(id: string, status: string): void {
    this._sendJson({ type: "task_status", id, status });
  }

  taskDone(id: string, result: string = ""): void {
    this._sendJson({ type: "task_done", id, result });
  }

  // --------- introspection ---------

  get connected(): boolean {
    return this.ws !== null && this.ws.readyState === 1;
  }

  get everConnected(): boolean {
    return this.connectedOnce;
  }
}

/**
 * Convenience: instantiate, open, and return a ready `MeshClient`.
 *
 * ```ts
 * const mesh = await connectMesh({ id: "cc7", name: "TS Bot", project: "OPTFINDER" });
 * mesh.on("message", e => console.log(`[${e.from}] ${e.text}`));
 * mesh.send("Hello from TypeScript");
 * ```
 */
export async function connectMesh(opts: ConnectOptions): Promise<MeshClient> {
  const client = new MeshClient(opts);
  await client.start();
  return client;
}

export default connectMesh;
